File: social.py
#!/usr/bin/python

class Illness:
    effort = 0.5
    time = 0 #发病结束倒计时
    cost = 0 #治疗花费
    death = 0.5
    posible = 0.1

class Person:
    "人"
    stat = {"kind":0, "work":0}
    age = 18.00
    XP = 0.01
    level = 0.5
    food = 0
    money = 0
    save = 0
    illness = Illness()
    @property
    def work_ability(self):
        pass
    @property
    def cost_ability(self):
        pass
    @property
    def happiness(self):
        pass

    def update(self):
        "日更新值"
        self.age += 0.01


class Worker(Person):
    def update(self):
        super().update()
        if self.stat["work"] == 1:
            self.XP += 0.001
        else if self.XP >= 0.001
            self.XP -= 0.002

    def have_work(self):
        self.stat["work"] = 1

    def lose_work(self):
        self.stat["work"] = 0

class Capital(Person):
    "资"
    worker = []
    wages = 100    #工资
    cupidity = 0.5 #贪心率
    @property
    def happiness(self):
        pass

    def __init__(self, worker = [Person()]):
        self.worker.append(worker)
    def add_worker(self):
        if len(self.worker)
        self.worker()

class Social:
    "社会环境"
    pop = []
    need_work = []
    cap = []
    price = 200
    food = 0
    require = 0

    def __init__(self, pop = 10, cap = 1, worker = [1], price = 200):
        if cap > len(worker) or sum(worker) > pop:
            print("Error:impossible number")
            return
        self.pop = [Worker() for i in range(pop)]
        self.need_work = self.pop
        self.cap = [Capital(self.pop[sum(worker[:i]):worker[i]]) for i in range(cap)]

if __name__ == "__main__":
    print("Here are nothing")

File: social.c
/*
 *   Copyright (C) 2024 u0_a70
 *
 *   文件名称：social.c
 *   创 建 者：u0_a70
 *   创建日期：2024年09月16日
 *   描    述：劳资模拟小程序
 *
 */


#include "include/tools.h"
#include <cstdlib>
#include <stdio.h>

struct social {
	int population;
	int capital;
	int prices;
	int food;
	int time;
} Social = { 10, 1, 200, 0, 0 };

struct worker {
	int flag;
	int age;
	int xp;
	int food;
	int money;
	double save;
	/* Note:
	 * work_ability
	 * cost_ability
	 * life_level
	 * happiness
	 * */
};

struct link {
	void *p;
	struct link *np;
};

struct capital {
	struct link *worker;
	int wages;	/* 工资 */
	int cupidity;	/* 贪心率 */
};


struct link *worker = NULL;
struct link *capital = NULL;


#define T_WORK (0b01)
int worker_flag(struct worker *p,int TYPE, int flag)
{
	if (!p) return -1;
	p->flag ^= (TYPE & (flag ? ~0 : 0));
	return (p->flag & TYPE ? 1 : 0);
}

int capital_add_worker(struct capital *cap)
{
	for (struct link *p = worker; p != NULL; p = p->np) {
		if (worker_flag(p->p, T_WORK, 0)) continue;
		worker_flag(p->p, T_WORK, 1);
		if (!cap->worker) {
			cap->worker = malloc(sizeof(struct link));
			cap->worker->p = p->p;
		}
	}
	return 0;
}

struct link *create_link(int len, void *context)
{
	struct link *head = NULL, *p = NULL, *lp = NULL;

	head = p = malloc(sizeof(struct link));
	for (int i = 0; i < len; i++) {
		p->p = context;
		lp = p;
		lp->np = p = malloc(sizeof(struct link));
	}
	lp->np = NULL;
	free(p);
	return head;
}

struct worker *create_worker()
{
	struct worker *p = malloc(sizeof(struct worker));
	p->flag = 0;
	p->age = 18;
	p->xp = 0;
	p->food = 0;
	p->money = 0;
	p->save = 0.5;
	return p;
}

struct capital *create_capital()
{
	struct capital *p = malloc(sizeof(struct worker));
	p->worker = NULL;
	p->wages = 200;
	p->cupidity = 1;
	return p;
}

int init()
{
	struct link *p = NULL, *lp = NULL;

	worker = create_link(Social.population, create_worker());
	capital = create_link(Social.capital, create_capital());
	return 0;
}

int run()
{
	for (int i = 0; i < 20; i++) {
		Social.time++;
		for (struct link *p = worker; p; p = p->np) {
			((struct worker*)p->p)->age += 20 % Social.time == 0 ? 1 : 0;
			((struct worker*)p->p)->xp += 20 % Social.time == 0 ? 1 : 0;
		}
	}
}

int main(int argc, char *argv[])
{
	int ch = 0;
	while ((ch = getopt(argc, argv, "hp:c:")) != -1) {	/* 获取参数 */
		switch (ch) {
		case '?':
		case 'h':
			// help();
			printf("Usage: xxx [-h] [-p <population>] [-c <capital>]\n");
			return 0;
			break;
		case 'p':
			Social.population = strtod(optarg, NULL);
			break;
		case 'c':
			Social.capital = strtod(optarg, NULL);
			break;
		default:
			break;
		}
	}
	init();
	printf("Pop:%d,Cap:%d\n", Social.population, Social.capital);
	capital_add_worker(capital->p);
	return 0;
}

